lib/open_api_spex/cast/discriminator.ex

```diff
@@ -0,0 +1,104 @@
+defmodule OpenApiSpex.Cast.Discriminator do
+  @moduledoc """
+  Defines the `OpenApiSpex.Discriminator.t` type.
+  """
+
+  alias OpenApiSpex.{Cast, Reference}
+
+  @enforce_keys :propertyName
+  defstruct [
+    :propertyName,
+    :mapping
+  ]
+
+  @typedoc """
+  [Discriminator Object](https://swagger.io/specification/#discriminatorObject)
+
+  When request bodies or response payloads may be one of a number of different schemas,
+  a discriminator object can be used to aid in serialization, deserialization, and validation.
+  The discriminator is a specific object in a schema which is used to inform the consumer of the
+  specification of an alternative schema based on the value associated with it.
+
+  A discriminator requires a composite key be set on the schema:
+
+    * `allOf`
+    * `oneOf`
+    * `anyOf`
+  """
+  @type t :: %__MODULE__{
+          propertyName: String.t(),
+          mapping: %{String.t() => %Reference{} | String.t()} | nil
+        }
+
+  def cast(ctx) do
+    case cast_discriminator(ctx) do
+      {:ok, result} -> {:ok, result}
+      error -> error
+    end
+  end
+
+  defp cast_discriminator(%_{value: value, schema: schema} = ctx) do
```

Comment:
This matches structs rather than plain maps. Basically, it ensures the map has a `:__struct__` key. The `_` is a throw-away variable that takes on the struct's module value.

I don't know if using it adds anything to this code, as it's assumed the struct is always a `%Cast{}` struct.

Reply:
Interesting. Seems like that just matches on any type of struct then?